### ISSUE TITLE
Unable to install Jaeger exporter in .NET Core 2.1 web app

### ISSUE THREAD

**Issue by MikeGoldsmith** [open-telemetry/opentelemetry-dotnet#605]
Raised by lingabalaji17 on Gitter.

Unable to install OpenTelemetry.Exporter.Jaeger (v0.2.0-alpha.242) nuget package from MyGet source into a default ASP.NET Core web app that is targeting .NET Core 2.1.

Severity | Code | Description | Project | File | Line | Suppression State
-- | -- | -- | -- | -- | -- | --
Error | NU1107 | Version conflict detected for Microsoft.AspNetCore.Http.Abstractions. Install/reference Microsoft.AspNetCore.Http.Abstractions 2.2.0 directly to project WebApplication2 to resolve this issue.  WebApplication2 -> OpenTelemetry.Exporter.Jaeger 0.2.0-alpha.242 -> ApacheThrift 0.13.0.1 -> Microsoft.AspNetCore.Http.Abstractions (>= 2.2.0)  WebApplication2 -> Microsoft.AspNetCore.App 2.1.1 -> Microsoft.AspNetCore.Http.Abstractions (>= 2.1.1 && < 2.2.0). | WebApplication2 | Z:\code\WebApplication2\WebApplication2.csproj | 1 |  

@CodeBlanch could this be related to your recent work?

**Comment by CodeBlanch** [comment]
@MikeGoldsmith Yes introduced when taking the official ApacheThrift NuGet on #494.
```
WebApplication2 -> OpenTelemetry.Exporter.Jaeger 0.2.0-alpha.242 -> ApacheThrift 0.13.0.1 -> Microsoft.AspNetCore.Http.Abstractions (>= 2.2.0)
WebApplication2 -> Microsoft.AspNetCore.App 2.1.1 -> Microsoft.AspNetCore.Http.Abstractions (>= 2.1.1 && < 2.2.0)
```
Looks like ApacheThrift 0.13.0.1 wants a newer version of Http.Abstraction than what 2.1 is targeting. I'll mess with it and see if I get it to work. Maybe we can downgrade the Thrift package when targeting netcoreapp 2.1?

**Comment by cijothomas** [comment]
@CodeBlanch I wonder how to target different version based on the runtime version? The exporter package is targetting netstandard2.0, and netstandard2.1. 

When using in asp.net core 2.1 or 2.2,, wont it always take the netstandard20 targets? I'd like to know if there is a workaround. 

Another library I own has the same issue..(library targets netstandard20, but depends on 2.2 version of a package, so it throws same nuget error when installing on asp.net core 2.1 project)

**Comment by CodeBlanch** [comment]
@cijothomas We might have to add a target for all versions of .NET Core (in addition to what we have for Standard) and then map them to the Thrift versions we need. Imagine something like this:

```
<TargetFrameworks>netstandard2.0;netstandard2.1;netcoreapp3.1;netcoreapp3.0;netcoreapp2.2;netcoreapp2.1;netcoreapp2.0</TargetFrameworks>
```
```
    <PackageReference Include="ApacheThrift" Version="0.13.0.1" IncludeAssets="all" ExcludeAssets="compile;runtime" GeneratePathProperty="true" Condition="'$(TargetFramework)' != 'netcoreapp2.1' and '$(TargetFramework)' != 'netcoreapp2.0'" />
    <PackageReference Include="ApacheThrift" Version="0.12.0.1" IncludeAssets="all" ExcludeAssets="compile;runtime" GeneratePathProperty="true" Condition="'$(TargetFramework)' == 'netcoreapp2.1' or '$(TargetFramework)' == 'netcoreapp2.0'" />
```
I took a stab at it but I wasn't able to test consuming via .NET Core 2.x/3.x because there are other issues. ApacheThrift 0.12.0.1 doesn't have the [netstd](https://github.com/apache/thrift/tree/0.13.0/lib/netstd) lib. We would have to consume the [netcore](https://github.com/apache/thrift/tree/0.13.0/lib/netcore) lib (12 or 13) BUT it has a different API. Namespace changes, TTransport doesn't exist, etc. So we're looking at a lot of preprocessor directives and ugliness to stitch it together.

Might be better to only support .NET Core 2.2+ for the Jaeger library?

**Comment by cijothomas** [comment]
mm. How about depending on 0.12.0.1 for NETSTANDARD20? SO all Asp.Net Core 2.0,2.1,2.2 runtime gets old version of Thrift which wont require the 2.2 packages?

_Might be better to only support .NET Core 2.2+ for the Jaeger library?_
Officially 2.2 has reached end of life. 2.1 is the LTS version. So only supporting 2.2+ is probably as good as saying, we only support .NET 3.0 or higher.

**Comment by CodeBlanch** [comment]
@cijothomas That's a good point! If we are OK with all the preprocessor directives I can take a stab at making it work targetting 0.12.0.1 for NETSTANDARD20, 0.13.0.1 for NETSTANDARD2.1.

```
    <TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
```
```
    <PackageReference Include="ApacheThrift" Version="0.13.0.1" IncludeAssets="all" ExcludeAssets="compile;runtime" GeneratePathProperty="true" Condition="'$(TargetFramework)' == 'netstandard2.1'" />
    <PackageReference Include="ApacheThrift" Version="0.12.0.1" IncludeAssets="all" ExcludeAssets="compile;runtime" GeneratePathProperty="true" Condition="'$(TargetFramework)' == 'netstandard2.0'" />
```

**Comment by cijothomas** [comment]
> @cijothomas That's a good point! If we are OK with all the preprocessor directives I can take a stab at making it work targetting 0.12.0.1 for NETSTANDARD20, 0.13.0.1 for NETSTANDARD2.1.
> 
> ```
>     <TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
> ```
> 
> ```
>     <PackageReference Include="ApacheThrift" Version="0.13.0.1" IncludeAssets="all" ExcludeAssets="compile;runtime" GeneratePathProperty="true" Condition="'$(TargetFramework)' == 'netstandard2.1'" />
>     <PackageReference Include="ApacheThrift" Version="0.12.0.1" IncludeAssets="all" ExcludeAssets="compile;runtime" GeneratePathProperty="true" Condition="'$(TargetFramework)' == 'netstandard2.0'" />
> ```

I think that's the right solution.
The Jaeger.Tests project is now running only .netcoreapp3.1, so we may want to add netcoreapp2.1 as well, since our Thrift versions differ in 2.xx and 3.xx

**Comment by MikeGoldsmith** [comment]
Thanks for looking into this @CodeBlanch @cijothomas :+1:

**Comment by CodeBlanch** [comment]
Ran into a problem, it looks like the 0.12.0.1 version isn't signed:
```
2>CSC : error CS8002: Referenced assembly 'Thrift, Version=0.12.0.0, Culture=neutral, PublicKeyToken=null' does not have a strong name.
```
Not sure where to go from here. I can put back a custom build of Thrift? The NuGet version is a real challenge to target.

**Comment by MikeGoldsmith** [comment]
So Thrift v0.12.0 is not signed but v0.13.0 is? That's super weird 👎

**Comment by cijothomas** [comment]
From the nuget description,  0.12.0.1 fixes the signing dll issue of 0.12.0,
@CodeBlanch Maybe clear you bin/obj directory and try again.

If still an issue, we can ask as an issue in Thrift Repo?

**Comment by cijothomas** [comment]
I downloaded the nuget locally and it looks signed! .12.0.1
![image](https://user-images.githubusercontent.com/5232798/79617400-4222b800-80bc-11ea-9543-8314cbdcb2c9.png)


Following is .12.0 which is clearly not signed.
![image](https://user-images.githubusercontent.com/5232798/79617461-641c3a80-80bc-11ea-8ffd-1d278f56da97.png)

**Comment by CodeBlanch** [comment]
@cijothomas Interesting! I downloaded the packages last night but didn't notice that digital signatures tab showing in Windows. Here's the assembly info shown in dotPeek:

![image](https://user-images.githubusercontent.com/28367120/79618755-d04c6d80-80bf-11ea-88f1-e1d321810585.png)

See how 0.13.0.1 has a PublicKeyToken and 0.12.0.1 does not? I'm not sure how the digital signature plays into that, but that's why it won't compile AFAIK.

I pushed my work up to a [branch](https://github.com/CodeBlanch/opentelemetry-dotnet/tree/jaeger-thrift-coreapp2x-issue) if you want to take it for a spin. I will also throw an Issue up on the Thrift repo right now although based on my previous contribution I don't think they'll be that interested in patching 12 for us. Their release process is intense.
